src/HexViewer.ts

```diff
@@ -291,16 +303,49 @@ export class HexViewer {
     }
 
     public deselect() {
-        this.setSelection(-1, -1);
+        this.setCursor(-1);
     }
 
-    public setSelection(start: number, end?: number) {
+    public setCursor (cursor: number) {
+        this.selectionCursor = cursor;
+        this.setSelection(
+            this.selectionCursor
+        );
+    }
+
+    public shiftCursor (cursor: number) {
+        var start = this.selectionStart;
+        var end = this.selectionEnd;
+        if (this.selectionCursor === end) {
+            end = cursor;
+        } else if (this.selectionCursor === start) {
+            start = cursor;
+        }
+        this.selectionCursor = cursor;
+        this.setSelection(
+            Math.max(0, start),
+            end
+        );
+    }
+
+    public setSelection(start: number, end?: number, cursor?: number) {
         if (this.isRecursive) return;
         end = end || start;
 
         var oldStart = this.selectionStart, oldEnd = this.selectionEnd;
-        this.selectionStart = start < end ? start : end;
-        this.selectionEnd = Math.min(start < end ? end : start, this.dataProvider.length - 1);
+        this.selectionStart = Math.min(start, end);
+        this.selectionEnd = Math.min(
+            Math.max(start, end),
+            this.dataProvider.length - 1
+        );
+
+        // is cursor out of bounds because called externally?
+        if (
+            this.selectionCursor < this.selectionStart
+            || this.selectionCursor > this.selectionEnd
+        ) cursor = this.selectionEnd;
+        if (cursor) this.selectionCursor = cursor;
```

Comment:
A weird condition, but in any case, this is the result of storing the `selectionCursor` separately. You wouldn't have to handle these states of inconsistency if you weren't storing the same data in different variables.